components/core/src/clp/streaming_compression/lzma/Compressor.cpp

```diff
@@ -0,0 +1,207 @@
+#include "Compressor.hpp"
+
+#include <array>
+#include <cstddef>
+#include <cstdint>
+#include <cstring>
+
+#include <lzma.h>
+#include <spdlog/spdlog.h>
+
+#include "../../ErrorCode.hpp"
+#include "../../FileWriter.hpp"
+#include "../../TraceableException.hpp"
+#include "../../type_utils.hpp"
+#include "Constants.hpp"
+
+namespace clp::streaming_compression::lzma {
+auto Compressor::init_lzma_encoder(lzma_stream* strm, int compression_level, size_t dict_size)
+        -> void {
+    lzma_options_lzma options;
+    if (0 != lzma_lzma_preset(&options, compression_level)) {
+        SPDLOG_ERROR("Failed to initialize LZMA options' compression level.");
+        throw OperationFailed(ErrorCode_BadParam, __FILENAME__, __LINE__);
+    }
+    options.dict_size = dict_size;
+    std::array<lzma_filter, 2> filters{{
+            {.id = LZMA_FILTER_LZMA2, .options = &options},
+            {.id = LZMA_VLI_UNKNOWN, .options = nullptr},
+    }};
+
+    // Initialize the encoder using a preset. Set the integrity to check
+    // to CRC64, which is the default in the xz command line tool. If
+    // the .xz file needs to be decompressed with XZ Embedded, use
+    // LZMA_CHECK_CRC32 instead.
+    auto const rc{lzma_stream_encoder(strm, filters.data(), LZMA_CHECK_CRC64)};
+
+    // Return successfully if the initialization went fine.
+    if (LZMA_OK == rc) {
+        return;
+    }
+
+    // Something went wrong. The possible errors are documented in
+    // lzma/container.h (src/liblzma/api/lzma/container.h in the source
+    // package or e.g. /usr/include/lzma/container.h depending on the
+    // install prefix).
+    char const* msg{nullptr};
+    switch (rc) {
+        case LZMA_MEM_ERROR:
+            msg = "Memory allocation failed";
+            break;
+
+        case LZMA_OPTIONS_ERROR:
+            msg = "Specified preset is not supported";
+            break;
+
+        case LZMA_UNSUPPORTED_CHECK:
+            msg = "Specified integrity check is not supported";
+            break;
+
+        default:
+            // This is most likely LZMA_PROG_ERROR indicating a bug in liblzma
+            msg = "Unknown error";
+            break;
+    }
+
+    SPDLOG_ERROR("Error initializing the encoder: {} (error code {})", msg, static_cast<int>(rc));
+    throw OperationFailed(ErrorCode_BadParam, __FILENAME__, __LINE__);
```

Comment:
The method is set to static because we want to use this OperationFailed.

two questions @kirkrodrigues :
1. Should we avoid static class member function if possible or is it ok to use it?
2. We can let this function to return a boolean, and let caller throw the operation? In that case, we can make this function to be a local function. The error message should still be output this function though

Reply:
1. In this case, I would avoid the static class member function since as you imply, it's not really necessary and it's preferable to keep the class' interface small.
2. `OperationFailed` is a public type, so the method can just throw it, right?

Reply:
Thanks for the fix idea.